### ISSUE TITLE
Consistent low recall on medical dataset

### ISSUE THREAD

**Issue by mw-boop** [ultralytics/ultralytics#20429]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone, 

I've been training the YOLO-v11s-seg on a medical dataset. After trying out a lot and after just finishing a grid search, I still get an incredibly low sensitivity. My specificity and Dice scores are alright. First things first:

**DATASET**
Dataset: train/val split: 
Train set: 4480 images, 3364 healthy, 1116 tumor
Val: 4804 images, 2859 healthy, 1945 tumor. 

The validation set was hand picked to include all sorts of different tumors: small, big, hard/easy to segment by hand. 

What is good to know is that by default EVERY frame has only ONE tumor. 

All frames went through the exact same preprocessing (normalization) and are the same size/recorded with the exact same system parameters.

**TRAINING ARGS**

task: segment
mode: train
model: yolo11s-seg.pt
epochs: 150
time: null
patience: 500
batch: 32
imgsz: 384
save: true
save_period: -1
cache: false
device: cuda
workers: 10
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: true
rect: false
cos_lr: false
close_mosaic: 0
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.1
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.4
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: true
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: null
nms: false
lr0: 0.04
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 3.44
cls: 4.16
dfl: 2.02
pose: 12.0
kobj: 1.0
nbs: 64
hsv_h: 0.0
hsv_s: 0.0
hsv_v: 0.0
degrees: 20
translate: 0.2
scale: 0.3
shear: 20
perspective: 0.0
flipud: 0.0
fliplr: 0.5
bgr: 0.0
mosaic: 0.0
mixup: 0.0
copy_paste: 0.0
copy_paste_mode: flip
auto_augment: null
erasing: 0.0
crop_fraction: 1.0
cfg: null
tracker: botsort.yaml

**INFERENCE ARGS (rest is default)**
Setting,Value
model_path,\\weights\best.pt
imgsz,384
task,segment
device,cuda
batch,256
save_conf,True
max_det,1
save_txt,True
save,True

**Grid search results**
My latest grid search differed in loss configs/augmentation profiles/image sizes:
# Augmentation profiles
AUGMENTATION_PROFILES = {
    "light": dict(degrees=10, translate=0.1, scale=0.2, shear=10, fliplr=0.5),
    "medium": dict(degrees=20, translate=0.2, scale=0.3, shear=20, fliplr=0.5),
    "heavy": dict(degrees=30, translate=0.5, scale=0.5, shear=30, fliplr=0.5)
}
# Loss configs
LOSS_CONFIGS = {
    "default": dict(box=7.5, cls=0.5, dfl=1.5),         # Official YOLOv8 default weights
    "seg-boost": dict(box=3.44, cls=4.16, dfl=2.02)      # Custom segmentation-boosted weights
}

# Values to grid search
IMG_SIZES = [256, 384, 512]

These are the results 

![Image](https://github.com/user-attachments/assets/1f613fcf-8b21-482f-ab6d-0ad9874198be)

And here are the results from yolo training from the 'best performing' of this bunch: seg-boost loss function, 'heavy' augmentation and 256 images ize. 

![Image](https://github.com/user-attachments/assets/7cb41321-8124-4add-852b-db75897449ab)

![Image](https://github.com/user-attachments/assets/93508538-d21e-42b3-a4e6-df32f4516266)

![Image](https://github.com/user-attachments/assets/40cae7b4-191e-4415-b6a6-bf1d04f4d49a)

Are there settings I'm missing? Anything that stands out to you?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mw-boop, thank you for sharing these detailed results and for your interest in Ultralytics 🚀! We appreciate the comprehensive information on your dataset, training configuration, and grid search efforts.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the issue more efficiently.

If this is a custom training ❓ Question about improving recall or segmentation performance, please include as much information as possible, such as sample images (if possible), training logs, or any additional details that might help us understand your setup. Also, ensure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For quick guidance, check out the [Docs](https://docs.ultralytics.com/) where you’ll find examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, and answers to many common questions.

Join the Ultralytics community to connect with others and get real-time support:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussion
- [Subreddit](https://reddit.com/r/Ultralytics) for community threads

## Upgrade

Please make sure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure your issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can run in any of these verified environments (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

✨ This is an automated response to help get you started. An Ultralytics engineer will review your issue and assist you soon!

**Comment by Y-T-G** [comment]
Instead of posting all the possible training arguments, you should only post what was modified. Otherwise this is hard to read and understand what was changed.

**Comment by Y-T-G** [comment]
> All frames went through the exact same preprocessing (normalization)

Do you mean you manually normalized it? The preprocessing is handled by Ultralytics. You shouldn't be normalizing it manually.

**Comment by Y-T-G** [comment]
You shouldn't be normalizing manually and it probably makes the augmentations. Ultralytics expects non-normalized images.

**Comment by mw-boop** [comment]
I ran again but now without normalizing the images, unfortunately, still the same low sensitivity.

**Comment by Y-T-G** [comment]
How large are your objects?

**Comment by mw-boop** [comment]
It depends, here is my tumor area distribution in pixels, original image size of all objects is 752x820

![Image](https://github.com/user-attachments/assets/cb37b8e8-a929-469d-ad01-836d626e79de)

**Comment by Y-T-G** [comment]
Can you post your labels.jpg?

**Comment by mw-boop** [comment]
![Image](https://github.com/user-attachments/assets/4a0af1bf-be19-41e1-9e30-e14b9aa1b6c3)

**Comment by glenn-jocher** [comment]
Based on the information provided and looking at your tumor distribution histogram and labels image, I can see several potential causes for your low recall issue:

1. **Small object detection challenge**: Many of your tumors appear quite small relative to the image size. YOLO models typically struggle with small objects, especially in medical imaging where context and subtle features are critical.

2. **Class imbalance**: Your training set has fewer tumor images (1116) compared to healthy ones (3364), which could bias the model toward predicting "no tumor."

Here are some specific suggestions to improve recall:

- **Try image tiling**: Instead of resizing the full 752x820 images down to 384px, consider tiling your larger images into smaller segments that maintain the original resolution of small tumors. This approach is mentioned in our [medical imaging blog](https://www.ultralytics.com/blog/using-yolo11-for-tumor-detection-in-medical-imaging) as particularly effective for tumor detection.

- **Adjust IoU threshold**: Your current IoU of 0.4 may be too strict for small objects. Try lowering it to 0.25-0.3 during training.

- **Balance class weights**: Consider using class weights or sampling techniques to address the imbalance in your dataset.

- **Increase anchor density**: For medical imaging with small objects, you might benefit from more anchors focused on smaller scales.

Have you explored any of these approaches? I'd be particularly interested to see if image tiling helps with your recall metrics.

**Comment by Y-T-G** [comment]
I didn't notice this before (which is why I said not to simply dump all possible arguments because it makes it hard to read). 

> Train set: 4480 images, 3364 healthy, 1116 tumor
> Val: 4804 images, 2859 healthy, 1945 tumor.

Why are you using `single_cls` if you have two classes?

**Comment by mw-boop** [comment]
Hi,


I see this in my inbox:
"
I didn't notice this before (which is why I said not to simply dump all possible arguments because it makes it hard to read).

Train set: 4480 images, 3364 healthy, 1116 tumor
Val: 4804 images, 2859 healthy, 1945 tumor.

Why are you using single_cls if you have two classes?"



However, I cannot find it now in the thread.

The reason I'm using single_cls is because there is only one class to detect (tumor) and it's either there or not, or is that not the way I should go about it?


________________________________
Van: Mohammed Yasin ***@***.***>
Verzonden: vrijdag 2 mei 2025 03:18
Aan: ultralytics/ultralytics ***@***.***>
CC: Maurijn Wieringa ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [ultralytics/ultralytics] Consistent low recall on medical dataset (Issue #20429)

[https://avatars.githubusercontent.com/u/32206511?s=20&v=4]Y-T-G left a comment (ultralytics/ultralytics#20429)<https://github.com/ultralytics/ultralytics/issues/20429#issuecomment-2846045760>

I didn't notice this before (which is why I said not to simply dump all possible arguments because it makes it hard to read).

Train set: 4480 images, 3364 healthy, 1116 tumor
Val: 4804 images, 2859 healthy, 1945 tumor.

Why are you using single_cls if you have two classes?

**Comment by Y-T-G** [comment]
How many classes do you have in the YAML?

**Comment by mw-boop** [comment]
Just one! 
names:
- tumor
path: xxx
train: images/train
val: images/val

**Comment by Y-T-G** [comment]
`single_cls` wouldn't make a difference in that case.

**Comment by mw-boop** [comment]
One thing i've noticed is that while running inference, changing the confidence threshold to 0.0001 makes the sensitivity high 
0.8098 and specificity low 0.5607, which is logical. But when I up my confidence threshold a little bit to 0.002, it already lowers to .68 sens and 0.86 spec. Of course, this is logical behaviour, however, what I think is not logical is that it's at such minuscule thresholds..

**Comment by glenn-jocher** [comment]
@mw-boop The extremely low confidence threshold situation is quite revealing - it suggests your model is detecting tumors but with very low confidence. This is common with class imbalance (3:1 healthy to tumor ratio in training) and small objects.

A few targeted suggestions:

1. **Adjust warmup_epochs**: Try setting `warmup_epochs=0` to prevent the learning rate from starting too low, as mentioned in our [model evaluation guide](https://docs.ultralytics.com/guides/model-evaluation-insights/).

2. **Modify training loss weights**: Increase the `cls` weight further from your current 4.16 to something like 8.0 to push the model to be more confident in classifications.

3. **Try focal loss**: Add `fl_gamma=2.0` to your training arguments, which helps focus learning on hard-to-classify examples.

4. **Multi-scale training**: Enable `multi_scale=True` since your tumors vary greatly in size.

The fact that your model performs better on the training set also suggests some overfitting. Consider adding more dropout (try 0.2 or 0.3) and implementing more aggressive augmentation specifically for the tumor class.

**Comment by mw-boop** [comment]
I've adjusted this and it worked @glenn-jocher! 
At conf=0.01 in inference, these are the results when it comes to confidence scores. 

Validation Set:                                                                                                           
  Confidence score range: 0.0101 – 0.8440
  Mean confidence score:   0.3415
  Median confidence score: 0.2191
Training Set:
  Confidence score range: 0.0102 – 0.8967
  Mean confidence score:   0.6621
  Median confidence score: 0.7707

                  DICE                 SENS              SPEC
Train:         0.8965,             0.9982,         0.9923
Validation: 0.7412,             0.5553,         0.9685

It's still not very confident, but I think that after trying out all what was said above, some grid search experiments with loss functions, augmentation settings and image sizes, that the only thing left to really try is figuring out the class imbalance. Luckily i'm still collecting new data, so hopefully more data will up the sensitivity. Thank you all for your help. Here are the latest results from my run. Any last tips are greatly appreciated, let me know if there is a patreon link or some other way i can express my gratitude in. 

Training settings i've changed:
Image size 256, augmentations degrees=10, translate=0.1, scale=0.2, shear=10, fliplr=0.5, cls loss= 8, dropout=0.1, iou=0.3

Inference settings i've changed:
conf=0.01, max_det=1. 

Can it be interesting to put iou in inference at 0.3 instead of the default 0.7 as well? 


![Image](https://github.com/user-attachments/assets/4e99c602-a680-47bb-9426-d416322e04fd)

**Comment by glenn-jocher** [comment]
@mw-boop I'm glad to see my suggestions helped improve your results! Looking at your updated metrics, I notice there's still a substantial gap between training and validation performance (training SENS: 0.9982 vs validation SENS: 0.5553), which indicates some overfitting.

Regarding your questions:

1. **Yes, lowering the IoU during inference to 0.3** could help with detecting small tumors. Since you've already lowered your training IoU to 0.3, it makes sense to match this during inference.

2. **Additional suggestions to reduce overfitting and improve validation sensitivity:**
   - Balance your dataset further by using weighted sampling or augmenting the minority class
   - Implement early stopping with the `patience` parameter set lower (try 20-30 epochs)
   - Consider implementing image tiling as mentioned earlier, which works especially well for small medical objects

Your current approach of collecting more data is absolutely the right direction - more diverse tumor examples will help the model generalize better.

There's no Patreon for Ultralytics, but if you'd like to contribute, consider sharing your work as a case study or contributing to the Ultralytics repository when your project is complete. The community benefits greatly from real-world medical imaging applications!
